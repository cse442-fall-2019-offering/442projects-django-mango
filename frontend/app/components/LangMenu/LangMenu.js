import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getLanguages } from 'my-actions/languagesActions';
import { makeSelectLanguages } from 'my-selectors/languagesSelectors';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import MultiSelect from '@khanacademy/react-multi-select';

function createOptions(list) {
  const output = [];
  list.forEach(function(element) {
    output.push({ label: element, value: element });
  });
  return output;
}

class LangMenu extends Component {
  state = {
    languages: [],
  };

  static getDerivedStateFromProps(props, prevState) {
    if (props.languages.length < 1) {
      return {
        loading: true,
      };
    }
    if (prevState.languages.length === 0) {
      return {
        languages: createOptions(props.languages),
        loading: false,
      };
    }
    return null;
  }

  componentDidMount() {
    const { onGetLanguages } = this.props;
    onGetLanguages();
  }

  render() {
    if (this.state.loading) {
      return null;
    }
    const { selectedLanguages, onLanguagesChange } = this.props;
    const { languages } = this.state;

    return (
      <MultiSelect
        options={languages}
        selected={selectedLanguages}
        onSelectedChanged={selected => onLanguagesChange(selected)}
        overrideStrings={{
          selectSomeItems: 'Programming Languages',
          allItemsAreSelected: 'All Languages are Selected',
          selectAll: 'Select All',
          search: 'Search',
        }}
      />
    );
  }
}

LangMenu.propTypes = {
  onGetLanguages: PropTypes.func.isRequired,
  selectedLanguages: PropTypes.array.isRequired,
  onLanguagesChange: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  languages: makeSelectLanguages(),
});

const mapDispatchToProps = dispatch => ({
  onGetLanguages: () => dispatch(getLanguages()),
});

const LangMenuMapped = connect(
  mapStateToProps,
  mapDispatchToProps,
)(LangMenu);

export default LangMenuMapped;
