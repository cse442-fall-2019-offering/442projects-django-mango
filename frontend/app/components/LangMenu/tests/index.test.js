import React from 'react';
import renderer from 'react-test-renderer';
import LangMenu from '../LangMenu';

test('LangMenu Component', () => {
  const component = renderer.create(
    <LangMenu
      onLanguagesChange={value => value}
      selectedLanguagegs={['python', 'javascript', 'c++']}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
