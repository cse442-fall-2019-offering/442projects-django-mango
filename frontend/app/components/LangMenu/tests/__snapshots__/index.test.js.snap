// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMediumEditor Component 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>Test</p>",
    }
  }
  onLanguagesChange={[Function]}
  selectedLanguages={[Array]}
/>
`;
